File: content/snippets/go/interview_lcci_03_01_three_in_one.go
package main

import "fmt"

type TripleInOne struct {
	size int
	data []int
	tops [3]int
}

func NewTripleInOne(stackSize int) *TripleInOne {
	return &TripleInOne{
		size: stackSize,
		data: make([]int, 3*stackSize),
	}
}

func (t *TripleInOne) Push(stackNum int, value int) {
	if t.tops[stackNum] == t.size {
		return
	}
	index := stackNum*t.size + t.tops[stackNum]
	t.data[index] = value
	t.tops[stackNum]++
}

func (t *TripleInOne) Pop(stackNum int) int {
	if t.tops[stackNum] == 0 {
		return -1
	}
	t.tops[stackNum]--
	index := stackNum*t.size + t.tops[stackNum]
	return t.data[index]
}

func (t *TripleInOne) Peek(stackNum int) int {
	if t.tops[stackNum] == 0 {
		return -1
	}
	index := stackNum*t.size + t.tops[stackNum] - 1
	return t.data[index]
}

func (t *TripleInOne) IsEmpty(stackNum int) bool {
	return t.tops[stackNum] == 0
}

func main() {
	s := NewTripleInOne(2)
	s.Push(0, 10)
	s.Push(0, 11)
	s.Push(0, 12)
	s.Push(1, 20)
	fmt.Println(s.Peek(0))
	fmt.Println(s.Pop(0))
	fmt.Println(s.Pop(0))
	fmt.Println(s.Pop(0))
	fmt.Println(s.IsEmpty(1))
}

File: content/snippets/python/interview_lcci_03_01_three_in_one.py
class TripleInOne:
    def __init__(self, stack_size: int) -> None:
        self.size = stack_size
        self.data = [0] * (3 * stack_size)
        self.tops = [0, 0, 0]

    def push(self, stack_num: int, value: int) -> None:
        if self.tops[stack_num] == self.size:
            return
        index = stack_num * self.size + self.tops[stack_num]
        self.data[index] = value
        self.tops[stack_num] += 1

    def pop(self, stack_num: int) -> int:
        if self.tops[stack_num] == 0:
            return -1
        self.tops[stack_num] -= 1
        index = stack_num * self.size + self.tops[stack_num]
        return self.data[index]

    def peek(self, stack_num: int) -> int:
        if self.tops[stack_num] == 0:
            return -1
        index = stack_num * self.size + self.tops[stack_num] - 1
        return self.data[index]

    def is_empty(self, stack_num: int) -> bool:
        return self.tops[stack_num] == 0


def main() -> None:
    stacks = TripleInOne(2)
    stacks.push(0, 10)
    stacks.push(0, 11)
    stacks.push(0, 12)
    stacks.push(1, 20)
    print(stacks.peek(0))
    print(stacks.pop(0))
    print(stacks.pop(0))
    print(stacks.pop(0))
    print(stacks.is_empty(1))


if __name__ == "__main__":
    main()
